### PR TITLE
fixes mem0 with valkey/redis issue

### PR DESCRIPTION
**Pull request by MatthiasHowellYopp**
changes to support using mem0 with valkey/redis as external memory provider for crewai

fixes #4031 

---

> [!NOTE]
> **Medium Risk**
> Changes request shaping for Mem0 search filters with provider-specific behavior, which could alter query results for Valkey/Redis users; test coverage is expanded to reduce regressions.
> 
> **Overview**
> Fixes `Mem0Storage` search filters for local Mem0 vector stores that don’t support nested `AND`/`OR` filter structures (Valkey/Redis) by **flattening single-condition `AND` filters** to a plain dict, while **leaving multi-condition/`OR` filters and cloud `MemoryClient` behavior unchanged**.
> 
> Adds targeted tests covering Valkey/Redis flattening, non-flattening cases, `None` config edge cases, and the cloud-vs-local guard; also updates pytest setup to reset the last event id between tests for better isolation.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 7a255c90101f7cab5f16c9f7522848f87d4ab36e. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: lib/crewai/tests/storage/test_mem0_storage.py [hunk 1]
```diff
@@ -384,6 +384,49 @@ def test_search_method_with_memory_oss(mem0_storage_with_mocked_config):
...
+        mem0_storage.memory.search.assert_called_once_with(
+            query="test query",
+            limit=5,
+            user_id="test_user",
+            filters={"run_id": "my_run_id"},  # Flattened from {"AND": [{"run_id": "my_run_id"}]}
+            threshold=0.5,
+        )
+
+        assert len(results) == 2
+        assert results[0]["content"] == "Result 1"
```
[the remark this answers is not in the document]
Comment (MatthiasHowellYopp) [reply lib/crewai/tests/storage/test_mem0_storage.py:429]: See test_magicmock_with_spec_passes_isinstance_check in the test_mem0_storage.py for proof that MagicMock(spec=x) does pass the isinstance test.

**Comment by Vidit-Ostwal** [comment]
Hey @MatthiasHowellYopp, maybe you can collapse all the commits into one for a cleaner git history

Try this
```
git checkout feature-branch
git reset --soft main
git commit -m "Add <feature-name>"
git push --force-with-lease
```

I am not a maintainer, feel free to ignore this...

**Comment by greysonlalonde** [comment]
> Hey @MatthiasHowellYopp, maybe you can collapse all the commits into one for a cleaner git history
> 
> Try this
> 
> ```
> git checkout feature-branch
> git reset --soft main
> git commit -m "Add <feature-name>"
> git push --force-with-lease
> ```
> 
> I am not a maintainer, feel free to ignore this...

This would be great!

**Comment by MatthiasHowellYopp** [comment]
@greysonlalonde changes done, looks like I need you (or someone) to let the checks run.

**Comment by MatthiasHowellYopp** [comment]
@greysonlalonde can you review - or recommend someone who can? tks.

**Comment by MatthiasHowellYopp** [comment]
@greysonlalonde Can we try again? looks lIke I need workflow approval and a review.

**Comment by MatthiasHowellYopp** [comment]
@greysonlalonde Is there something I can do to get this to the other side of merged?  Seems I'm back to needing the checks permitted.

**Comment by MatthiasHowellYopp** [comment]
@greysonlalonde Can you approve the awating workflows, please?

**Comment by MatthiasHowellYopp** [comment]
fixes https://github.com/crewAIInc/crewAI/issues/4030

**Comment by MatthiasHowellYopp** [comment]
no longer relevant

**Comment by Vidit-Ostwal** [comment]
> no longer relevant

**Comment by MatthiasHowellYopp** [comment]
working on a version that works with the new memory architecture.

**Comment by Vidit-Ostwal** [comment]
> working on a version that works with the new memory architecture.
